src/gui/usbipd_gui.rs

```diff
@@ -150,22 +137,120 @@ impl UsbipdGui {
     }
 
-    fn show_tray_menu(&self) {
+    fn show_menu_tray(self: &Rc<UsbipdGui>) {
+        if let Some(handler) = self.menu_tray_event_handler.borrow().as_ref() {
+            nwg::unbind_event_handler(handler);
+        }
+
+        let mut menu_tray = nwg::Menu::default();
+        nwg::Menu::builder()
+            .popup(true)
+            .parent(self.window.handle)
+            .build(&mut menu_tray)
+            .unwrap();
+
+        let devices = self
+            .connected_tab_content
+            .connected_devices
+            .borrow()
+            .iter()
+            .cloned()
+            .collect::<Vec<_>>();
+
+        let mut menu_items: Vec<(nwg::MenuItem, Rc<UsbDevice>)> = Vec::with_capacity(devices.len());
+        for device in devices {
+            let device_name = device.description.as_deref();
+            let vid_pid = device.vid_pid();
+            let description = device_name.map(|s| s.to_string()).unwrap_or(
+                vid_pid
+                    .clone()
+                    .unwrap_or_else(|| "Unknown Device".to_string()),
+            );
+
+            if device.is_bound() {
+                let menu_item = self
+                    .new_menu_item(menu_tray.handle, &description, device.is_attached())
+                    .unwrap();
+
+                menu_items.push((menu_item, Rc::new(device.clone())));
+            }
+        }
+
+        self.new_menu_separator(menu_tray.handle).unwrap();
+        let open_item = self.new_menu_item(menu_tray.handle, "Open", false).unwrap();
+        self.new_menu_separator(menu_tray.handle).unwrap();
+        let exit_item = self.new_menu_item(menu_tray.handle, "Exit", false).unwrap();
+
+        let rc_self_weak = Rc::downgrade(&self);
+        *self.menu_tray_event_handler.borrow_mut() = Some(nwg::full_bind_event_handler(
+            &self.window.handle,
+            move |evt, _evt_data, handle| {
+                if let Some(rc_self) = rc_self_weak.upgrade() {
+                    match evt {
+                        nwg::Event::OnMenuItemSelected => {
+                            if handle == open_item.handle {
+                                UsbipdGui::show(&rc_self);
+                            } else if handle == exit_item.handle {
+                                UsbipdGui::exit();
+                            } else {
+                                for (menu_item, device) in menu_items.iter() {
+                                    if handle == menu_item.handle {
+                                        if device.is_attached() {
+                                            device.detach().ok();
+                                        } else {
+                                            device.attach().ok();
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                        _ => (),
```

Comment:
You should add a handler case for "menu closed" (don't know the name of the event) that unbinds the event handler here instead of where you do it at the beginning of this function.

Reply:
That will be much better! This was quite ugly to deal with.

Reply:
I tried the menu_exit, and unfortunately it caused the other handlers to not work. When they are clicked on, an exit event occurs, so it's either that:

* exit event is handled before the selected event (handler gets unregistered first)
* or they are handled in different threads causing a race condition

I could look into the implementation to understand.

I've left the unbind_event_handler in the menu open for now.

Reply:
Right, I'm fine with leaving as is, as long as you add a comment in the code explaining what you said here, for documentation purposes. Something along the lines of 
> Unbinding of the menu handler is done here because the exit event is received before menu action events, therefore unbinding on the exit event would cause other actions to not be handled.